packages/sparql-language-server/src/SparqlLanguageServer.ts

```diff
@@ -85,10 +87,36 @@ export class SparqlLanguageServer extends AbstractLanguageServer<
   }
 
   handleUpdateCompletionData(update: SparqlCompletionData) {
+    // `relationshipCompletionItems` and `typeCompletionItems` must be updated
+    // in 2 different scenarios:
+    // #1 - namespaces provided after relationshipBindings or typeBindings
+    // #2 - namespaces provided before relationshipBindings or typeBindings
+    // Otherwise you can find yourself with 1, both or neither reflecting the
+    // namespace prefixes based on the order the updates are processed, which is
+    // indeterminate.
     if (update.namespaceMap) {
       this.namespaceMap = namespaceArrayToObj(update.namespaceMap);
+      if (!update.relationshipBindings && this.relationshipBindings) {
+        this.relationshipCompletionItems = this.buildCompletionItemsFromData(
+          this.namespaceMap,
+          this.relationshipBindings.map((binding) => ({
+            iri: binding.relationship.value,
+            count: binding.count.value,
+          }))
+        );
+      }
+      if (!update.typeBindings && this.typeBindings) {
+        this.typeCompletionItems = this.buildCompletionItemsFromData(
+          this.namespaceMap,
+          this.typeBindings.map((binding) => ({
+            iri: binding.type.value,
+            count: binding.count.value,
+          }))
+        );
+      }
```

Comment:
Instead of literally duplicating the logic like this, couldn't we just change the original conditions slightly, like this?

```
if (update.namespaceMap) {
  this.namespaceMap = namespaceArrayToObj(update.namespaceMap);
}
if (update.relationshipBindings || (update.namespaceMap && this.relationshipBindings)) {
  this.relationshipBindings = /* . . . other stuff here . . . */
}
if (update.typeBindings || (update.namespaceMap && this.typeBindings)) {
  this.typeBindings = /* . . . other stuff here . . . */
}
```

Alternatively, we could put the duplicated logic into method calls. I think I'd prefer either to literally just repeating.

Reply:
It's more complicated than that -- the conditions would look like:

Reply:
FWIW the logic only moves by doing this: we trade 2 method calls for 2 additional `||`s and move 2 nested `ifs` to additional conditions on existing `ifs`. The result is a few lines shorter.

```
+++ b/packages/sparql-language-server/src/SparqlLanguageServer.ts
@@ -96,40 +96,26 @@ export class SparqlLanguageServer extends AbstractLanguageServer<
     // indeterminate.
     if (update.namespaces) {
       this.namespaceMap = namespaceArrayToObj(update.namespaces);
-      if (!update.relationshipBindings && this.relationshipBindings) {
-        this.relationshipCompletionItems = this.buildCompletionItemsFromData(
-          this.namespaceMap,
-          this.relationshipBindings.map((binding) => ({
-            iri: binding.relationship.value,
-            count: binding.count.value,
-          }))
-        );
-      }
-      if (!update.typeBindings && this.typeBindings) {
-        this.typeCompletionItems = this.buildCompletionItemsFromData(
-          this.namespaceMap,
-          this.typeBindings.map((binding) => ({
-            iri: binding.type.value,
-            count: binding.count.value,
-          }))
-        );
-      }
     }
-    if (update.relationshipBindings) {
-      this.relationshipBindings = update.relationshipBindings;
+    if (
+      update.relationshipBindings ||
+      (update.namespaces && this.relationshipBindings)
+    ) {
+      this.relationshipBindings =
+        update.relationshipBindings || this.relationshipBindings;
       this.relationshipCompletionItems = this.buildCompletionItemsFromData(
         this.namespaceMap,
-        update.relationshipBindings.map((binding) => ({
+        this.relationshipBindings.map((binding) => ({
           iri: binding.relationship.value,
           count: binding.count.value,
         }))
       );
     }
-    if (update.typeBindings) {
-      this.typeBindings = update.typeBindings;
+    if (update.typeBindings || (update.namespaces && this.typeBindings)) {
+      this.typeBindings = update.typeBindings || this.typeBindings;
       this.typeCompletionItems = this.buildCompletionItemsFromData(
         this.namespaceMap,
-        update.typeBindings.map((binding) => ({
+        this.typeBindings.map((binding) => ({
           iri: binding.type.value,
           count: binding.count.value,
         }))
```